File: server/src/plugins/runtime/python/namespaces.py
from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class CoreManagerNamespaces:
    core_manager_subject: str
    core_manager_rpc: str


@dataclass(frozen=True, slots=True)
class DeviceManagerNamespaces:
    device_manager_subject: str
    device_manager_rpc: str


@dataclass(frozen=True, slots=True)
class FrameWorkerNamespaces:
    frame_worker_child_rpc: str
    frame_worker_child: str


@dataclass(frozen=True, slots=True)
class PluginNamespaces:
    plugin_device_manager_subject: str
    plugin_child_rpc: str
    plugin_child: str
    plugin_storage_rpc: str
    plugin_config_store_rpc: str


@dataclass(frozen=True, slots=True)
class CameraNamespaces:
    camera_subject: str
    camera_controller_rpc: str


@dataclass(frozen=True, slots=True)
class PluginCameraNamespaces:
    camera_interfaces_rpc: str
    camera_storage_rpc: str
    camera_impl_rpc: str


@dataclass(frozen=True, slots=True)
class PluginSensorNamespaces:
    sensor_storage_rpc: str


@dataclass(frozen=True, slots=True)
class SensorControllerNamespaces:
    sensor_subject: str
    sensor_rpc: str


@dataclass(frozen=True, slots=True)
class SensorEventNamespaces:
    sensor_subject: str


@dataclass(frozen=True, slots=True)
class SensorProviderNamespaces:
    sensor_rpc: str


@dataclass(frozen=True, slots=True)
class FrameWorkerDetectionNamespaces:
    detection_rpc: str


@dataclass(frozen=True, slots=True)
class DetectionEventNamespaces:
    detection_event_subject: str


@dataclass(frozen=True, slots=True)
class DiscoveryManagerNamespaces:
    discovery_manager_subject: str
    discovery_manager_rpc: str


@dataclass(frozen=True, slots=True)
class DownloadManagerNamespaces:
    download_manager_rpc: str


@dataclass(frozen=True, slots=True)
class NotificationManagerNamespaces:
    notifications_publish_subject: str


class NamespaceManager:
    @staticmethod
    def core_manager_namespaces() -> CoreManagerNamespaces:
        return CoreManagerNamespaces(
            core_manager_subject="coreManager.subscriber",
            core_manager_rpc="coreManager.rpc",
        )

    @staticmethod
    def device_manager_namespaces() -> DeviceManagerNamespaces:
        return DeviceManagerNamespaces(
            device_manager_subject="deviceManager.subscriber",
            device_manager_rpc="deviceManager.rpc",
        )

    @staticmethod
    def discovery_manager_namespaces() -> DiscoveryManagerNamespaces:
        return DiscoveryManagerNamespaces(
            discovery_manager_subject="discoveryManager.subscriber",
            discovery_manager_rpc="discoveryManager.rpc",
        )

    @staticmethod
    def download_manager_namespaces() -> DownloadManagerNamespaces:
        return DownloadManagerNamespaces(
            download_manager_rpc="downloadManager.rpc",
        )

    @staticmethod
    def plugin_file_serve_rpc(plugin_id: str) -> str:
        return f"plugin.{plugin_id}.fileserve.rpc"

    @staticmethod
    def notification_manager_namespaces() -> NotificationManagerNamespaces:
        return NotificationManagerNamespaces(
            notifications_publish_subject="notifications.publish",
        )

    @staticmethod
    def plugin_namespaces(plugin_id: str) -> PluginNamespaces:
        return PluginNamespaces(
            plugin_device_manager_subject=f"plugin.{plugin_id}.deviceManager.subscriber",
            plugin_child_rpc=f"plugin.{plugin_id}.child.rpc",
            plugin_child=f"plugin.{plugin_id}.child",
            plugin_storage_rpc=f"plugin.{plugin_id}.storage.rpc",
            plugin_config_store_rpc=f"plugin.{plugin_id}.configstore.rpc",
        )

    @staticmethod
    def camera_namespaces(camera_id: str) -> CameraNamespaces:
        return CameraNamespaces(
            camera_subject=f"camera.{camera_id}.subscriber",
            camera_controller_rpc=f"camera.{camera_id}.controller.rpc",
        )

    @staticmethod
    def frame_worker_namespaces(camera_id: str) -> FrameWorkerNamespaces:
        return FrameWorkerNamespaces(
            frame_worker_child_rpc=f"camera.{camera_id}.frameWorker.child.rpc",
            frame_worker_child=f"camera.{camera_id}.frameWorker.child",
        )

    @staticmethod
    def plugin_camera_namespaces(plugin_id: str, camera_id: str) -> PluginCameraNamespaces:
        return PluginCameraNamespaces(
            camera_interfaces_rpc=f"plugin.{plugin_id}.camera.{camera_id}.cameraInterfaces.rpc",
            camera_storage_rpc=f"plugin.{plugin_id}.camera.{camera_id}.cameraStorage.rpc",
            camera_impl_rpc=f"plugin.{plugin_id}.camera.{camera_id}.impl.rpc",
        )

    @staticmethod
    def plugin_sensor_namespaces(plugin_id: str, camera_id: str, sensor_id: str) -> PluginSensorNamespaces:
        return PluginSensorNamespaces(
            sensor_storage_rpc=f"plugin.{plugin_id}.camera.{camera_id}.sensor.{sensor_id}.storage.rpc",
        )

    @staticmethod
    def sensor_controller_namespaces(camera_id: str) -> SensorControllerNamespaces:
        return SensorControllerNamespaces(
            sensor_subject=f"camera.{camera_id}.sensors.subject",
            sensor_rpc=f"camera.{camera_id}.sensors.rpc",
        )

    @staticmethod
    def sensor_event_namespaces(camera_id: str, sensor_id: str) -> SensorEventNamespaces:
        return SensorEventNamespaces(
            sensor_subject=f"camera.{camera_id}.sensor.{sensor_id}.subject",
        )

    @staticmethod
    def sensor_provider_namespaces(
        plugin_id: str, camera_id: str, sensor_id: str
    ) -> SensorProviderNamespaces:
        return SensorProviderNamespaces(
            sensor_rpc=f"plugin.{plugin_id}.camera.{camera_id}.sensor.{sensor_id}.rpc",
        )

    @staticmethod
    def frame_worker_detection_namespaces(camera_id: str) -> FrameWorkerDetectionNamespaces:
        return FrameWorkerDetectionNamespaces(
            detection_rpc=f"camera.{camera_id}.frameWorker.detection.rpc",
        )

    @staticmethod
    def detection_event_namespaces(camera_id: str) -> DetectionEventNamespaces:
        return DetectionEventNamespaces(
            detection_event_subject=f"camera.{camera_id}.events.subject",
        )

    @staticmethod
    def sensor_controller_rpc(camera_id: str) -> str:
        return f"camera.{camera_id}.sensors.controller.rpc"

File: server/src/rpc/namespaces.ts
export interface CoreManagerNamespaces {
  coreManagerSubject: string;
  coreManagerRpc: string;
}

export interface DeviceManagerNamespaces {
  deviceManagerSubject: string;
  deviceManagerRpc: string;
}

export interface FrameWorkerNamespaces {
  frameWorkerChildRpc: string;
  frameWorkerChild: string;
}

export interface PluginNamespaces {
  pluginDeviceManagerSubject: string;
  pluginChildRpc: string;
  pluginChild: string;
  pluginStorageRpc: string;
  pluginConfigStoreRpc: string;
}

export interface CameraNamespaces {
  cameraSubject: string;
  cameraControllerRpc: string;
}

export interface PluginCameraNamespaces {
  cameraInterfacesRpc: string;
  cameraStorageRpc: string;
  cameraImplRpc: string;
}

export interface PluginSensorNamespaces {
  sensorStorageRpc: string;
}

export interface SensorRegistryNamespaces {
  sensorsSubject: string;
  sensorsRpc: string;
}

export interface SensorCameraViewNamespaces {
  sensorWriteSubject: string;
}

export interface SensorEventNamespaces {
  sensorSubject: string;
}

export interface SensorProviderNamespaces {
  sensorRpc: string;
}

export interface FrameWorkerDetectionNamespaces {
  detectionRpc: string;
}

export interface DetectionEventNamespaces {
  detectionEventSubject: string;
}

export interface TerminalManagerNamespaces {
  terminalManagerSubject: string;
  terminalManagerRpc: string;
}

export interface SensorDiscoveryNamespaces {
  sensorDiscoverySubject: string;
}

export interface DiscoveryManagerNamespaces {
  discoveryManagerSubject: string;
  discoveryManagerRpc: string;
}

export interface DownloadManagerNamespaces {
  downloadManagerRpc: string;
}

export interface NotificationManagerNamespaces {
  notificationsPublishSubject: string;
}

export class NamespaceManager {
  static coreManagerNamespaces(): CoreManagerNamespaces {
    return {
      coreManagerSubject: 'coreManager.subscriber',
      coreManagerRpc: 'coreManager.rpc',
    };
  }

  static deviceManagerNamespaces(): DeviceManagerNamespaces {
    return {
      deviceManagerSubject: 'deviceManager.subscriber',
      deviceManagerRpc: 'deviceManager.rpc',
    };
  }

  static sensorDiscoveryNamespaces(): SensorDiscoveryNamespaces {
    return {
      sensorDiscoverySubject: 'sensorDiscovery.subscriber',
    };
  }

  static discoveryManagerNamespaces(): DiscoveryManagerNamespaces {
    return {
      discoveryManagerSubject: 'discoveryManager.subscriber',
      discoveryManagerRpc: 'discoveryManager.rpc',
    };
  }

  static pluginNamespaces(pluginId: string): PluginNamespaces {
    return {
      pluginDeviceManagerSubject: `plugin.${pluginId}.deviceManager.subscriber`,
      pluginChildRpc: `plugin.${pluginId}.child.rpc`,
      pluginChild: `plugin.${pluginId}.child`,
      pluginStorageRpc: `plugin.${pluginId}.storage.rpc`,
      pluginConfigStoreRpc: `plugin.${pluginId}.configstore.rpc`,
    };
  }

  static pluginFileServeRpc(pluginId: string): string {
    return `plugin.${pluginId}.fileserve.rpc`;
  }

  static cameraNamespaces(cameraId: string): CameraNamespaces {
    return {
      cameraSubject: `camera.${cameraId}.subscriber`,
      cameraControllerRpc: `camera.${cameraId}.controller.rpc`,
    };
  }

  static frameWorkerNamespaces(cameraId: string): FrameWorkerNamespaces {
    return {
      frameWorkerChildRpc: `camera.${cameraId}.frameWorker.child.rpc`,
      frameWorkerChild: `camera.${cameraId}.frameWorker.child`,
    };
  }

  static pluginCameraNamespaces(pluginId: string, cameraId: string): PluginCameraNamespaces {
    return {
      cameraInterfacesRpc: `plugin.${pluginId}.camera.${cameraId}.cameraInterfaces.rpc`,
      cameraStorageRpc: `plugin.${pluginId}.camera.${cameraId}.cameraStorage.rpc`,
      cameraImplRpc: `plugin.${pluginId}.camera.${cameraId}.impl.rpc`,
    };
  }

  static pluginSensorNamespaces(pluginId: string, sensorId: string): PluginSensorNamespaces {
    return {
      sensorStorageRpc: `plugin.${pluginId}.sensor.${sensorId}.storage.rpc`,
    };
  }

  static sensorRegistryNamespaces(): SensorRegistryNamespaces {
    return {
      sensorsSubject: 'sensors.subject',
      sensorsRpc: 'sensors.rpc',
    };
  }

  static sensorCameraViewNamespaces(cameraId: string): SensorCameraViewNamespaces {
    return {
      sensorWriteSubject: `camera.${cameraId}.sensors.writes`,
    };
  }

  static sensorEventNamespaces(sensorId: string): SensorEventNamespaces {
    return {
      sensorSubject: `sensor.${sensorId}.subject`,
    };
  }

  static sensorProviderNamespaces(pluginId: string, sensorId: string): SensorProviderNamespaces {
    return {
      sensorRpc: `plugin.${pluginId}.sensor.${sensorId}.rpc`,
    };
  }

  static frameWorkerDetectionNamespaces(cameraId: string): FrameWorkerDetectionNamespaces {
    return {
      detectionRpc: `camera.${cameraId}.frameWorker.detection.rpc`,
    };
  }

  static detectionEventNamespaces(cameraId: string): DetectionEventNamespaces {
    return {
      detectionEventSubject: `camera.${cameraId}.events.subject`,
    };
  }

  static terminalManagerNamespaces(): TerminalManagerNamespaces {
    return {
      terminalManagerSubject: 'terminalManager.subscriber',
      terminalManagerRpc: 'terminalManager.rpc',
    };
  }

  static downloadManagerNamespaces(): DownloadManagerNamespaces {
    return {
      downloadManagerRpc: 'downloadManager.rpc',
    };
  }

  static notificationManagerNamespaces(): NotificationManagerNamespaces {
    return {
      notificationsPublishSubject: 'notifications.publish',
    };
  }

  static workerAgentRpc(agentId: string): string {
    return `worker.${agentId}.rpc`;
  }

  static workerManagerRpc(): string {
    return 'workers.manager.rpc';
  }

  static workerSync(agentId: string): string {
    return `worker.${agentId}.sync`;
  }

  static workerDisconnect(): string {
    return 'workers.disconnect';
  }

  static workerLogs(): string {
    return 'workers.logs';
  }
}
